### DISCUSSION TITLE
Dynamically modify Session Lifespan and update existing sessions.

### DISCUSSION THREAD

**Discussion by arieltorti** [ory/kratos#3153] [Q&A] [answered]
Hey !

We're running the self-hosted version of the Ory Kratos and we need the flexibility of modifying session lifespans. Is there a way to do this programmatically through an API [similar to what's provided when using Ory Network](https://www.ory.sh/docs/kratos/session-management/session-lifespan)?

Currently, the only option is to edit the schema files and restart the instances so the changes in the schemas take effect.

Moreover, would it be possible to bulk-update or bulk-expire user sessions when the lifespan is modified? After a few tests, I noticed that when the session lifespan is changed and the instances are restarted existing sessions are still active and under the effect of the old lifespan.

**Comment by aeneasr** [comment, marked as answer]
>Is there a way to do this programmatically through an API [similar to what's provided when using Ory Network](https://www.ory.sh/docs/kratos/session-management/session-lifespan)?

The functionality in the Ory Network is the same as in self-hosted for this feature. What you're probably looking for is: https://www.ory.sh/docs/kratos/session-management/refresh-extend-sessions

> Moreover, would it be possible to bulk-update or bulk-expire user sessions when the lifespan is modified? After a few tests, I noticed that when the session lifespan is changed and the instances are restarted existing sessions are still active and under the effect of the old lifespan.

I don't think we can add that - updating 100M+ rows on a config change can bring a system down :)

**Comment by arieltorti** [comment]
> The functionality in the Ory Network is the same as in self-hosted for this feature. What you're probably looking for is: https://www.ory.sh/docs/kratos/session-management/refresh-extend-sessions

That's similar but it's not what I'm looking for. I want to modify the session lifetime via an API, so that when I change it, for example, from 30 days -> 7 days newly created sessions expire in 7 days instead of 30.

In order to do that now I have to change Kratos config file and restart the service, I'm trying to avoid that because I need to change this programmatically.

> I don't think we can add that - updating 100M+ rows on a config change can bring a system down :)

Makes total sense.

**Reply by aeneasr** [reply]
> In order to do that now I have to change Kratos config file and restart the service, I'm trying to avoid that because I need to change this programmatically.

I see, API configuration is only available on the Ory Network. For open source, you'll need to rely on configuration files.

**Reply by arieltorti** [reply]
That's unfortunate, I'll use the container restarting approach for now.

Do you have plans on adding API configuration for self-hosted ? Or maybe an endpoint to trigger a configuration reload using the files ?

**Reply by aeneasr** [reply]
The configuration file is hotreloaded, so you don't need to restart the service. API based configuration is out of scope for self-hosted

**Reply by arieltorti** [reply]
One last detail, is there an endpoint to retrieve the current config with defaults applied? Similar to what was asked here: https://github.com/ory/kratos/discussions/3027

Basically, a read-only config API.

**Reply by vinckr** [reply]
No API when you are self-hosting, but since you have full control over that you probably can implement something that fits your use case. 
Would that also be for dynamically modifying session lifespans?

**Reply by arieltorti** [reply]
Exactly, if I modify the file Ory will hot-reload it but I want to be able to verify and query the new configuration. I can work around it but it would be helpful if Kratos provided a read config endpoint.

**Reply by arieltorti** [reply]
The read config endpoint is also useful for cases where values are overridden by environment variables.

**Comment by geekodour** [comment]
I've noticed that if I restart kratos, all user sessions are logged out and they need to re-log in. is this expected?
